docker/docker.go

```diff
@@ -150,8 +158,32 @@ func localImageExists(ctx context.Context, dClient *client.Client, hostname stri
 	return true, nil
 }
 
-func pullImage(ctx context.Context, dClient *client.Client, hostname string, containerImage string) error {
-	out, err := dClient.ImagePull(ctx, containerImage, types.ImagePullOptions{})
+func pullImage(ctx context.Context, dClient *client.Client, hostname string, containerImage string, prsMap map[string]v3.PrivateRegistry) error {
+
+	pullOptions := types.ImagePullOptions{}
+	containerNamed, err := ref.ParseNormalizedNamed(containerImage)
+	if err != nil {
+		return err
+	}
+	// We do this if we have some docker.io login information
+	if ref.Domain(containerNamed) == DockerRegistryURL {
```

Comment:
@moelsayed is there a reason why we distinguish between ref.Domain(containerNamed) being == vs != DockerRegistryURL? in both cases can we use `getRegistryAuth(pr)`? Or if there is a big diff between `pullOptions.RegistryAuth` vs `pullOptions.PrivilegeFunc`, can we refactor the code as follows, so it's a bit easier to read:

```
if pr, ok := prsMap[regURL]; ok {
  if regURL {
   // getRegistryAuth
  } else {
   //  tryRegistryAuth
   }
}
```

Reply:
Internally, `pullOptions.PrivilegeFunc` will generate and set `pullOptions.RegistryAuth` if the initial unauthenticated pull request failed with an Unauthorized error and try again.

However, docker registry API returns a not found error if the image is private, which doesn't trigger the `pullOptions.PrivilegeFunc` call. I am not sure why they do this. Security maybe ?

So, if it's a Docker registry, it could be a private or a public image. I check if I have auth credentials for docker. If I found it, I explicitly generate  `pullOptions.RegistryAuth` and use it directly.